Jupyter notebook support

MarcoGorelli opened the pull request:
closes #2345 

To summarise, based on what was discussed in that issue:
- due to not being able to parse automagics (e.g. `pip install black`) without a running IPython kernel, cells with syntax which is parseable by neither `ast.parse` nor `IPython` will be skipped
- cells with multiline magics will be skipped
- trailing semicolons will be preserved, as they are often put there intentionally in Jupyter Notebooks to suppress unnecessary output

----

some TODOs:
- [x] add black-primer test (e.g. pymc-examples)
- [x] support `--diff`
- [x] handle when explicitly trying to run on Jupyter notebook but extra dependencies aren't found

MarcoGorelli commented:
`black-primer` failed in CI in the previous run, but it passes for me locally:
```
$ black-primer
[2021-07-04 09:12:08,875] INFO: 20 projects to run Black over (lib.py:355)
[2021-07-04 09:12:09,083] INFO: Finished STDIN (lib.py:327)
[2021-07-04 09:12:09,768] INFO: Finished aioexabgp (lib.py:327)
[2021-07-04 09:12:11,684] INFO: Finished bandersnatch (lib.py:327)
[2021-07-04 09:12:11,907] INFO: Finished attrs (lib.py:327)
[2021-07-04 09:12:12,867] INFO: Finished channels (lib.py:327)
[2021-07-04 09:12:14,020] INFO: Finished flake8-bugbear (lib.py:327)
[2021-07-04 09:12:25,082] INFO: Finished hypothesis (lib.py:327)
[2021-07-04 09:13:44,317] INFO: Finished pandas (lib.py:327)
[2021-07-04 09:13:53,801] INFO: Finished django (lib.py:327)
[2021-07-04 09:13:58,901] INFO: Finished pillow (lib.py:327)
[2021-07-04 09:14:00,742] INFO: Finished ptr (lib.py:327)
[2021-07-04 09:14:02,053] INFO: Finished poetry (lib.py:327)
[2021-07-04 09:14:03,596] INFO: Finished pyanalyze (lib.py:327)
[2021-07-04 09:14:12,242] INFO: Finished pyramid (lib.py:327)
[2021-07-04 09:14:14,876] INFO: Finished pytest (lib.py:327)
[2021-07-04 09:14:20,426] INFO: Finished tox (lib.py:327)
[2021-07-04 09:14:42,349] INFO: Finished sqlalchemy (lib.py:327)
[2021-07-04 09:14:46,559] INFO: Finished typeshed (lib.py:327)
[2021-07-04 09:14:46,637] INFO: Finished virtualenv (lib.py:327)
[2021-07-04 09:15:03,724] INFO: Finished warehouse (lib.py:327)
[2021-07-04 09:15:03,724] INFO: Analyzing results (lib.py:379)
-- primer results 📊 --

20 / 20 succeeded (100.0%) ✅
0 / 20 FAILED (0.0%) 💩
 - 0 projects disabled by config
 - 0 projects skipped due to Python version
 - 0 skipped due to long checkout
```

Any ideas what the issue might be?

MarcoGorelli commented:
Example of diff:

![image](https://user-images.githubusercontent.com/33491632/124832339-6fcf7780-df74-11eb-908d-3adf7371ef24.png)

JelleZijlstra commented on src/black/__init__.py:
```diff
@@ -825,6 +853,9 @@ def format_file_contents(src_contents: str, *, fast: bool, mode: Mode) -> FileCo
     `mode` is passed to :func:`format_str`.
     """
+    if mode.is_ipynb:
+        return format_ipynb_string(src_contents, mode=mode)
```
This means we're skipping the equivalent and stable checks below for notebooks.

MarcoGorelli replied:
Yes, you're right thanks. I've changed it so that, unless `--fast` is passed, the checks will be performed on a cell-by-cell basis, but after the cells have been masked (I don't see how to do it with the raw cells because they might not be parseable by `ast.parse` if they contain magics)

JelleZijlstra commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,388 @@
+    assert magic
+    nbytes = max(len(magic) // 2 - 1, 1)
+    token = secrets.token_hex(nbytes)
+    while token in src:  # pragma: nocover
```
Can you make this throw an exception after say 100 iterations? Just so we don't go into an infinite loop if token_hex is badly broken.

JelleZijlstra commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,388 @@
+            for arg in node.value.args:
+                assert isinstance(arg, ast.Str)
+                args.append(arg.s)
+            header: Optional[str] = f"%%{args[0]}"
```
Why do you need the annotation? The local variable is never None.

JelleZijlstra commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,388 @@
+    and we look for instances of the latter.
+    """
+
+    def __init__(self) -> None:
```
Prefer making this a dataclass

JelleZijlstra commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,388 @@
+
+    def __init__(self) -> None:
+        """Record where magics occur."""
+        self.magics: Dict[int, List[Tuple[int, str]]] = collections.defaultdict(list)
```
Same, let's use a dataclass

MarcoGorelli replied:
Just to clarify - which part would you like to be a dataclass? I've done it for the `Tuple[int, str]`, but did you mean it for the whole thing? If so, it's not clear to me how to make a dataclass out of a `collections.defaultdict`, but I'll think about what I can do here

JelleZijlstra commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,388 @@
+        and we look for instances of any of the latter.
+        """
+        if isinstance(node.value, ast.Call) and _is_ipython_magic(node.value.func):
+            assert isinstance(node.value.func, ast.Attribute)  # help mypy
```
Using a PEP 647 TypeGuard should make this unnecessary. Mypy doesn't support that very well yet though, so no worries if you're not interested.

MarcoGorelli replied:
That's awesome, thanks!

Would this require typing_extensions as a runtime dependency before Python 3.10?

ichard26 replied:
We already require typing_extensions for `Final` < Python 3.8 so don't worry :)

MarcoGorelli replied:
Cool, done - TypeGuard's really neat, it's been worth it trying to contribute to `black` even if just to learn about TypeGuard's existence!

JelleZijlstra commented on tests/test_ipynb.py:
```diff
@@ -0,0 +1,286 @@
+@pytest.mark.parametrize(
+    "src, expected",
+    (
+        pytest.param("ls =!ls", "ls = !ls", id="System assignment"),
```
What if the input already contains `get_ipython().system` or similar? The output should still be `get_ipython().system`, not something transformed back into magics.

MarcoGorelli replied:
That's an excellent point, thanks - looks like they can't be distinguished after `TransformerManager` has transformed the cell:
```
>>> TransformerManager().transform_cell("get_ipython().system('ls')")
"get_ipython().system('ls')\n"
>>> TransformerManager().transform_cell("!ls")
"get_ipython().system('ls')\n"
```

The only solution I can think of is ignore cells which contain these transformed magics

JelleZijlstra reviewed (changes requested):
Thank you for putting in the work here!

Also, this should be mentioned in the docs somewhere, not just in README.

MarcoGorelli commented:
Hey, thanks for your excellent review, really appreciate it!

I've tried addressing all points, here's a screenshot of the part I've added to the docs:

![image](https://user-images.githubusercontent.com/33491632/125162298-ea6be300-e17e-11eb-8433-cc022d198671.png)

terasakisatoshi commented:
Hi, @MarcoGorelli thank you for your great work.
I have a question. If this PR is merged, can we format a paired markdown file, say `sample.md` as below via `jupytext` so that keeps trailing semicolons will be preserved?

![image](https://user-images.githubusercontent.com/16760547/125240720-6e5fcf80-e325-11eb-88ae-b4df659e6a34.png)

The current `black`  formatter

```python
>>> black.__version__
'21.6b0'
```

will delete semicolons `;` after `plt.plot(z);` or `z = [x, y];` Indeed, `diff` will appear if we run a command as follow:

```
jupytext --pipe 'black {} --diff' sample.md
[jupytext] Reading sample.md in format md
[jupytext] Executing black /var/folders/14/cg438fvd28dg9wpvpwt1lz0c0000gn/T/sample_849gg7za.py --diff
--- /var/folders/14/cg438fvd28dg9wpvpwt1lz0c0000gn/T/sample_849gg7za.py 2021-07-12 06:31:15.338389 +0000
+++ /var/folders/14/cg438fvd28dg9wpvpwt1lz0c0000gn/T/sample_849gg7za.py 2021-07-12 06:31:15.500610 +0000
@@ -17,9 +17,9 @@
 from matplotlib import pyplot as plt

 # %%
 x = 1
 y = 2
-z = [x, y];
+z = [x, y]

 # %%
-plt.plot(z);
+plt.plot(z)
would reformat /var/folders/14/cg438fvd28dg9wpvpwt1lz0c0000gn/T/sample_849gg7za.py
All done! ✨ 🍰 ✨
1 file would be reformatted.
[jupytext] Updating the timestamp of sample.md
```

I want to keep these semicolons.

MarcoGorelli commented:
Hi @terasakisatoshi 

If you want to try out this branch, you could do
```
pip uninstall black -y
pip install git+https://github.com/MarcoGorelli/black.git@jupyter
pip install -U IPython tokenize-rt
```
If you have lots of notebooks, it would be really useful if you could try this out and report back on any bugs

Currently you need to pass an `.ipynb` file, or as least something that could be parsed by `json.dumps`, so I don't think it would work to pass an `.md` file via `jupytext`.

If this PR is accepted, however, then perhaps formatting input "ipynb-style" could be a follow-up PR

terasakisatoshi commented:
> Hi @terasakisatoshi
> 
> If you want to try out this branch, you could do
> 
> ```
> pip uninstall black -y
> pip install git+https://github.com/MarcoGorelli/black.git@jupyter
> pip install -U IPython tokenize-rt typing-extensions
> ```
> 
> If you have lots of notebooks, it would be really useful if you could try this out and report back on any bugs
> 

Thank you for your comment. It seems my `.ipynb` files are formatted as if [jupyterlab-code-formatter](https://jupyterlab-code-formatter.readthedocs.io/en/latest/) does. That's so cool.

> Currently you need to pass an `.ipynb` file, or as least something that could be parsed by `json.dumps`, so I don't think it would work to pass an `.md` file via `jupytext`.

I see. Since I manage notebooks with `.md` files, I'm trying to find a good way to format files with a code formatter (e.g. `black` here) that keeps trailing semicolons `plt.plot(blahblah);`.

mwouts commented:
Hi @MarcoGorelli , it's great to see you working on this!

Well to comment on the workflow involving `black` and `jupytext`, here is a short overview of what one can use:
```sh
# convert ipynb to py:percent, apply black, convert back to ipynb & preserve outputs
jupytext --pipe black sample.ipynb

# convert md to py:percent, apply black, convert back to md
jupytext --pipe black sample.md

# (with your PR) convert md to ipynb, apply black, convert back to md
jupytext --pipe black --pipe-fmt ipynb sample.md
```

I have no clue unfortunately on how to preserve the trailing semicolons.

MarcoGorelli commented:
> I have no clue unfortunately on how to preserve the trailing semicolons.

This PR addresses that, so the suggestion above to run `black` on the `ipynb` file directly will work:

```
% cat trailing_semicolon.md
---
jupyter:
  jupytext:
    cell_metadata_filter: -all
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.11.3
  kernelspec:
    display_name: Python 3
    language: python
    name: python3
---

```python
fig, ax = plt.subplots()
ax.plot(xdata,ydata);
```
% black trailing_semicolon.ipynb 
reformatted trailing_semicolon.ipynb
All done! ✨ 🍰 ✨
1 file reformatted.
(good-bayesian) marcogorelli@OVMG025 tmp % jupytext --sync trailing_semicolon.ipynb 
[jupytext] Reading trailing_semicolon.ipynb in format ipynb
[jupytext] Updating trailing_semicolon.md
% cat trailing_semicolon.md
---
jupyter:
  jupytext:
    cell_metadata_filter: -all
    formats: ipynb,md
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.11.3
  kernelspec:
    display_name: Python 3
    language: python
    name: python3
---

```python
fig, ax = plt.subplots()
ax.plot(xdata, ydata);
```
```

MarcoGorelli commented on src/black/handle_ipynb_magics.py:
```diff
@@ -0,0 +1,422 @@
+    body: str
+
+
+@dataclasses.dataclass
+class CellMagicFinder(ast.NodeVisitor):
+    """Find cell magics.
+
+    Note that the source of the abstract syntax tree
+    will already have been processed by IPython's
+    TransformerManager().transform_cell.
+
+    For example,
+
+        %%time\nfoo()
+
+    would have been transformed to
+
+        get_ipython().run_cell_magic('time', '', 'foo()\\n')
+
+    and we look for instances of the latter.
+    """
+
+    cell_magic: Optional[CellMagic] = None
```
> Same, let's use a dataclass

Done - is this what you meant @JelleZijlstra ? (sorry having taken some time to address that, I wasn't aware that dataclasses would inherit from other classes - there, that's another new thing I've learned from this PR 💚  )

ichard26 replied:
Annnddd it looks like this broke compatibility with mypyc:

```console
~/programming/oss/black on mypyc-support-pt2 [$?⇡] via Python v3.8.5 (bm-venv) took 1s445ms 
❯ black setup.py --check -v
Traceback (most recent call last):
  File "/home/ichard26/programming/oss/black/bm-venv/bin/black", line 5, in <module>
    from black import patched_main
  File "src/black/__init__.py", line 47, in <module>
    from black.files import find_project_root, find_pyproject_toml, parse_pyproject_toml
  File "/home/ichard26/programming/oss/black/bm-venv/lib/python3.8/site-packages/black/files.py", line 26, in <module>
    from black.handle_ipynb_magics import jupyter_dependencies_are_installed
  File "src/black/handle_ipynb_magics.py", line 316, in <module>
    class CellMagicFinder(ast.NodeVisitor):
AttributeError: attribute '__dict__' of 'type' objects is not writable
```

(it's broken on main too but I'm currently working on mypyc integration on the `mypyc-support-pt2` branch). I'll deal with this myself since I need to assess how badly this broke things. The good news is that this should be easy to work around (hopefully just remove the dataclass decorator) since I've seen and dealt with issues like this before.

cooperlees commented on setup.py:
```diff
@@ -79,14 +79,15 @@ def get_long_description() -> str:
         "colorama": ["colorama>=0.4.3"],
         "python2": ["typed-ast>=1.4.2"],
         "uvloop": ["uvloop>=0.15.2"],
+        "jupyter": ["ipython>=7.8.0", "tokenize-rt>=3.2.0"],
```
Since I see we're using dataclasses in black code now, should we explicitly add it as a dep for when we're < 3.7 ?

JelleZijlstra commented on CHANGES.md:
```diff
@@ -9,6 +9,7 @@
 - Add primer support and test for code piped into black via STDIN (#2315)
 - Fix internal error when `FORCE_OPTIONAL_PARENTHESES` feature is enabled (#2332)
 - Accept empty stdin (#2346)
+- Add support for Jupyter Notebook (#2357)
```
This needs to be moved

JelleZijlstra commented on README.md:
```diff
@@ -41,7 +41,8 @@ Try it out now using the [Black Playground](https://black.vercel.app). Watch the
 _Black_ can be installed by running `pip install black`. It requires Python 3.6.2+ to
 run. If you want to format Python 2 code as well, install with
-`pip install black[python2]`.
+`pip install black[python2]`. If you want format Jupyter Notebooks as well, install with
```
```suggestion
`pip install black[python2]`. If you want to format Jupyter Notebooks, install with
```

JelleZijlstra commented on docs/getting_started.md:
```diff
@@ -18,7 +18,8 @@ Also, you can try out _Black_ online for minimal fuss on the
 _Black_ can be installed by running `pip install black`. It requires Python 3.6.2+ to
 run, but can format Python 2 code too. Python 2 support needs the `typed_ast`
-dependency, which be installed with `pip install black[python2]`.
+dependency, which be installed with `pip install black[python2]`. If you want format
+Jupyter Notebooks as well, install with `pip install black[jupyter]`.
```
```suggestion
dependency, which be installed with `pip install black[python2]`. If you want to format
Jupyter Notebooks, install with `pip install black[jupyter]`.
```

JelleZijlstra commented:
Trying it out locally I found a bug in the caching logic:
- I ran it once without installing `[jupyter]` and got a message telling me to install it
- I `pip install -e .[jupyter]`
- But now I get `...ipynb wasn't modified on disk since last run`

To fix this we need to make sure we don't write ipynb files to the cache if we don't actually format them.

ichard26 commented on pyproject.toml:
```diff
@@ -31,4 +31,5 @@ build-backend = "setuptools.build_meta"
 optional-tests = [
   "no_python2: run when `python2` extra NOT installed",
   "no_blackd: run when `d` extra NOT installed",
+  "no_jupyter: run when `jupyter` extra NOT installed",
```
So a thing with us maintainers is that we like to run bare `pytest $args` without using tox (mostly for speed and flexibility) and currently the test suite runs but then all of the Jupyter related tests crash with an ImportError as expected. It would be better if either a) test collection would fail immediately if the Jupyter dependencies aren't available, or alternatively b) juptyer tests could be automatically skipped (but that's unideal since it would be better to test as much as possible within a single `pytest` invocation).

For A, see what happens when the `d` extras aren't installed.

MarcoGorelli replied:
That's a good point, thanks. I've addressed this so that `tests/test_ipynb` will not run if the extra dependencies aren't installed

ichard26 commented on src/black/__init__.py:
```diff
@@ -196,6 +206,14 @@ def validate_regex(
         " when piping source on standard input)."
     ),
 )
+@click.option(
+    "--ipynb",
+    is_flag=True,
+    help=(
+        "Format all input files like ipynb notebooks regardless of file extension "
+        "(useful when piping source on standard input)."
+    ),
+)
```
I'm not the biggest fan of this since this allows for "illegal state" to be configured (forcing all files to be treated both as stubs and jupyter notebooks). While this doesn't break anything (the --ipynb flag wins out), it's probably not the best UX. Totally down to deal with this in a later PR (although before this gets released). Also, totally fine to leave this alone, it's not that terrible anyway.

MarcoGorelli replied:
Yeah, that's a good point...I'll remove it for now then

ichard26 commented on src/black/__init__.py:
```diff
@@ -741,6 +765,12 @@ def format_file_in_place(
         dst_contents = format_file_contents(src_contents, fast=fast, mode=mode)
     except NothingChanged:
         return False
+    except ModuleNotFoundError:
+        warnings.warn(
+            f"Skipping '{src}' as Jupyter dependencies are not installed.\n"
+            "You can fix this by running ``pip install black[jupyter]``"
+        )
+        return False
```
I don't like very much since it's kinda tacked on and could easily become misleading if we ever add additional lazy / conditional imports. It's also not immediately clear that this situation must be about missing Jupyter dependencies.

JelleZijlstra replied:
This is also what's causing the caching bug I noted above.

MarcoGorelli replied:
Excellent points here, thanks for catching this!

ichard26 commented on src/black/__init__.py:
```diff
@@ -829,26 +879,116 @@ def format_file_contents(src_contents: str, *, fast: bool, mode: Mode) -> FileCo
+    """
+    trailing_newline = src_contents[-1] == "\n"
+    modified = False
+    nb = json.loads(src_contents)
```
Maybe it's worth failing with a more descriptive error message when JSON decoding fails? Currently the JSONDecodeError is caught untouched.

MarcoGorelli replied:
Good point - added, thanks!

ichard26 reviewed (commented):
I have no experience Jupyter so I can't comment on the Jupyter specific code, but there are a few more minor issues I noted.

Thank you so much for the work, it's great seeing this support being implemented :tada:

ichard26 commented on src/black/__init__.py:
```diff
@@ -819,6 +852,23 @@ def format_stdin_to_stdout(
         f.detach()
 
 
+def check_src_and_dst_equivalent(
```
actually, this name isn't 100% clear because this also includes the stability check, I can't think of a good name tho.

MarcoGorelli replied:
how about `check_stability_and_equivalence`?

MarcoGorelli commented:
Thanks @JelleZijlstra and @ichard26 for your excellent reviews!

The biggest point was about the caching bug, which I've addressed by not processing `.ipynb` files if the optional dependencies aren't installed. I've added a test for this in `tests/test_ipynb.py`.

Nonetheless, to demo this from the terminal:

<details>

```console
(black) marco@marco-Predator-PH315-52:~/black-dev$ pip uninstall ipython -y
Found existing installation: ipython 7.25.0
Uninstalling ipython-7.25.0:
  Successfully uninstalled ipython-7.25.0
(black) marco@marco-Predator-PH315-52:~/black-dev$ black ../nbQA-dev/tests/data/notebook_for_testing.ipynb 
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
No Python files are present to be formatted. Nothing to do 😴
(black) marco@marco-Predator-PH315-52:~/black-dev$ pip install ipython
Collecting ipython
  Using cached ipython-7.25.0-py3-none-any.whl (786 kB)
Requirement already satisfied: traitlets>=4.2 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (5.0.5)
Requirement already satisfied: backcall in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.2.0)
Requirement already satisfied: matplotlib-inline in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.1.2)
Requirement already satisfied: setuptools>=18.5 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (57.0.0)
Requirement already satisfied: jedi>=0.16 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.18.0)
Requirement already satisfied: pygments in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (2.9.0)
Requirement already satisfied: decorator in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (5.0.9)
Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (3.0.19)
Requirement already satisfied: pexpect>4.3 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (4.8.0)
Requirement already satisfied: pickleshare in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.7.5)
Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from jedi>=0.16->ipython) (0.8.2)
Requirement already satisfied: ptyprocess>=0.5 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from pexpect>4.3->ipython) (0.7.0)
Requirement already satisfied: wcwidth in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython) (0.2.5)
Requirement already satisfied: ipython-genutils in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from traitlets>=4.2->ipython) (0.2.0)
Installing collected packages: ipython
Successfully installed ipython-7.25.0
WARNING: You are using pip version 21.0.1; however, version 21.1.3 is available.
You should consider upgrading via the '/home/marco/virtualenvs/black/bin/python -m pip install --upgrade pip' command.
(black) marco@marco-Predator-PH315-52:~/black-dev$ black ../nbQA-dev/tests/data/notebook_for_testing.ipynb 
reformatted ../nbQA-dev/tests/data/notebook_for_testing.ipynb
All done! ✨ 🍰 ✨
1 file reformatted.
(black) marco@marco-Predator-PH315-52:~/black-dev$ pip uninstall ipython -y
Found existing installation: ipython 7.25.0
Uninstalling ipython-7.25.0:
  Successfully uninstalled ipython-7.25.0
(black) marco@marco-Predator-PH315-52:~/black-dev$ black ../nbQA-dev/tests/data
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install black[jupyter]``
All done! ✨ 🍰 ✨
2 files left unchanged.
(black) marco@marco-Predator-PH315-52:~/black-dev$ pip install ipython
Collecting ipython
  Using cached ipython-7.25.0-py3-none-any.whl (786 kB)
Requirement already satisfied: decorator in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (5.0.9)
Requirement already satisfied: pexpect>4.3 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (4.8.0)
Requirement already satisfied: matplotlib-inline in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.1.2)
Requirement already satisfied: setuptools>=18.5 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (57.0.0)
Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (3.0.19)
Requirement already satisfied: jedi>=0.16 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.18.0)
Requirement already satisfied: pygments in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (2.9.0)
Requirement already satisfied: traitlets>=4.2 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (5.0.5)
Requirement already satisfied: backcall in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.2.0)
Requirement already satisfied: pickleshare in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from ipython) (0.7.5)
Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from jedi>=0.16->ipython) (0.8.2)
Requirement already satisfied: ptyprocess>=0.5 in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from pexpect>4.3->ipython) (0.7.0)
Requirement already satisfied: wcwidth in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->ipython) (0.2.5)
Requirement already satisfied: ipython-genutils in /home/marco/virtualenvs/black/lib/python3.8/site-packages (from traitlets>=4.2->ipython) (0.2.0)
Installing collected packages: ipython
Successfully installed ipython-7.25.0
WARNING: You are using pip version 21.0.1; however, version 21.1.3 is available.
You should consider upgrading via the '/home/marco/virtualenvs/black/bin/python -m pip install --upgrade pip' command.
(black) marco@marco-Predator-PH315-52:~/black-dev$ black ../nbQA-dev/tests/data
reformatted ../nbQA-dev/tests/data/clean_notebook_with_multiline.ipynb
reformatted ../nbQA-dev/tests/data/default_magic.ipynb
reformatted ../nbQA-dev/tests/data/invalid_import_in_doctest.ipynb
reformatted ../nbQA-dev/tests/data/comment_after_trailing_semicolon.ipynb
reformatted ../nbQA-dev/tests/data/non_default_magic.ipynb
reformatted ../nbQA-dev/tests/data/commented_out_magic.ipynb
reformatted ../nbQA-dev/tests/data/notebook_with_trailing_semicolon.ipynb
reformatted ../nbQA-dev/tests/data/percent_format.ipynb
reformatted ../nbQA-dev/tests/data/starting_with_comment.ipynb
reformatted ../nbQA-dev/tests/data/notebook_for_testing_copy.ipynb
reformatted ../nbQA-dev/tests/data/notebook_starting_with_md.ipynb
reformatted ../nbQA-dev/tests/data/notebook_with_indented_magics.ipynb
All done! ✨ 🍰 ✨
12 files reformatted, 15 files left unchanged.
```

</details>

ichard26 commented on tests/test_ipynb.py:
```diff
@@ -0,0 +1,422 @@
+    assert "reformatted" in result.output
+
+
+def test_cache_isnt_written_if_no_jupyter_deps_many(
```
This isn't firing `reformat_many` as the name suggests since there's still ultimately only one file discovered. 
https://github.com/psf/black/blob/8ea641eed5b9540287a8e9a9afa1458b72b9b630/src/black/__init__.py#L432-L447

This does hit the `get_python_files` code path though so maybe that's what you were going for?

MarcoGorelli replied:
Ah, what I meant was that it's running on a directory rather than on a single file - I've renamed it for clarity (as with the one below), thanks for pointing this out!

ichard26 commented on tests/test_no_ipynb.py:
```diff
@@ -0,0 +1,37 @@
+    assert expected_output in result.output
+
+
+def test_ipynb_diff_with_no_change_many(tmpdir: tmpdir) -> None:
```
Ditto with the similar test in `tests/test_ipynb.py`.

BotchJet-Pony commented:
Thanks for that, I will give a look later. I will give my best.

Em sex., 30 de jul. de 2021 16:42, Richard Si ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/black/handle_ipynb_magics.py
> <https://github.com/psf/black/pull/2357#discussion_r680180887>:
>
> > ***@***.***
>
> +class CellMagicFinder(ast.NodeVisitor):
>
> +    """Find cell magics.
>
> +
>
> +    Note that the source of the abstract syntax tree
>
> +    will already have been processed by IPython's
>
> +    TransformerManager().transform_cell.
>
> +
>
> +    For example,
>
> +
>
> +        %%time\nfoo()
>
> +
>
> +    would have been transformed to
>
> +
>
> +        get_ipython().run_cell_magic('time', '', 'foo()\\n')
>
> +
>
> +    and we look for instances of the latter.
>
> +    """
>
> +
>
> +    cell_magic: Optional[CellMagic] = None
>
>
> Annnddd it looks like this broke compatibility with mypyc:
>
> ~/programming/oss/black on mypyc-support-pt2 [$?⇡] via Python v3.8.5 (bm-venv) took 1s445ms
>
> ❯ black setup.py --check -v
> Traceback (most recent call last):
>   File "/home/ichard26/programming/oss/black/bm-venv/bin/black", line 5, in <module>
>     from black import patched_main
>   File "src/black/__init__.py", line 47, in <module>
>     from black.files import find_project_root, find_pyproject_toml, parse_pyproject_toml
>   File "/home/ichard26/programming/oss/black/bm-venv/lib/python3.8/site-packages/black/files.py", line 26, in <module>
>     from black.handle_ipynb_magics import jupyter_dependencies_are_installed
>   File "src/black/handle_ipynb_magics.py", line 316, in <module>
>     class CellMagicFinder(ast.NodeVisitor):
> AttributeError: attribute '__dict__' of 'type' objects is not writable
>
> (it's broken on main too but I'm currently working on mypyc integration on
> the mypyc-support-pt2 branch). I'll deal with this myself since I need to
> assess how badly this broke things. The good news is that this should be
> easy to work around (hopefully just remove the dataclass decorator) since
> I've seen and dealt with issues like this before.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/psf/black/pull/2357#discussion_r680180887>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARZEIIWCZJWD3RYMZHMPY73T2L6DTANCNFSM47YLKDDA>
> .
>

ichard26 reviewed (changes requested):
Overall it looks good to me, there's some possible cleaning up to do, but pretty all large PRs has deal with that. The only main change I'd like to see is to bring back the `--ipynb` flag, I thought about it some more and discussed it with Jelle and ended deciding / realizing that the ability to force ipynb mode is more important than the UX or internal representation (perhaps it will be fine in the end and it ends up just being a quirk of the codebase :slightly_smiling_face:). Sorry for backtracking on this and making you do useless work!

Thank you for pushing forward on this! Happy to see this usecase support.

ichard26 reviewed (approved):
SGTM, let's ship it (we can always deal with any remaining issues as they get reported).

ichard26 commented:
congratulations @MarcoGorelli and thank you very much for your persistence! happy to see this land :tada:
